rclcpp/doc/notes_on_statically_typed_parameters.md

```diff
@@ -0,0 +1,111 @@
+```
+
+```py
+rcl_interfaces.msg.ParameterDescriptor descriptor;
+descriptor.dynamic_typing = True;
+
+node.declare_parameter("dynamically_typed_parameter", None, descriptor);
+```
+
+## How is the parameter type specified?
+
+The parameter type will be inferred from the default value provided when declaring it.
+
+## Statically typed parameters when allowing undeclared parameters
+
+When undeclared parameters are allowed, the descriptor of undeclared parameters will have the dynamic typing field set.
+That's because in that case there is no declaration, and it doesn't make much sense to enforce the type of the first value set.
+
+TBD: Should parameters that were declared with the dynamic typing option off enforce that the parameter doesn't change the type even if allow undeclared parameters is set?
+In the current implementation that isn't enforce, but it might make sense to do it for parameters that were declared.
+
+## Declaring a parameter without a default value
+
+There might be cases were a default value does not make sense and the user must always provide an override.
+In those cases, the parameter type can be specified explicitly:
+
+```cpp
+// method signature
+template<typename T>
+Node::declare_parameter<T>(std::string name, rcl_interfaces::msg::ParameterDescriptor = rcl_interfaces::msg::ParameterDescriptor{});
+// or alternatively
+Node::declare_parameter(std::string name, rclcpp::ParameterType type, rcl_interfaces::msg::ParameterDescriptor = rcl_interfaces::msg::ParameterDescriptor{});
+
+// examples
+node->declare_paramter<int64_t>("my_integer_parameter");  // declare an integer parameter
+node->declare_paramter("another_integer_parameter", rclcpp::ParameterType::PARAMETER_INTEGER);  // another way to do the same
+```
+
+```py
+# method signature
+Node.declare_parameter(name: str, param_type: rclpy.Parameter.Type, descriptor: rcl_interfaces.msg.ParameterDescriptor = rcl_interfaces.msg.ParameterDescriptor())
+
+# example
+node.declare_paramter('my_integer_parameter', rclpy.Parameter.Type.INTEGER);  # declare an integer parameter
+```
+
+If the parameter is optional, e.g.: only needed depending on the value of another parameter, the recommended approach is to conditionally declare the parameter:
+
+```cpp
+auto mode = node->declare_parameter("mode", "modeA");  // "mode" parameter is an string
+if (mode == "modeB") {
+    node->declare_parameter<int64_t>("param_needed_when_mode_b");  // when "modeB", the user must provide "param_needed_when_mode_b"
+}
+```
+
+## Other migration notes
+
+Declaring a parameter with only a name is deprecated:
+
+```cpp
+node->declare_parameter("my_param");  // this generates a build warning
+```
+
+```py
+node.declare_parameter("my_param");  # this generates a python user warning
+```
+
+Before, you could initialize a parameter with the "NOT SET" value (in cpp `rclcpp::ParameterValue{}`, in python `None`).
+Now this will throw an exception in both cases:
```

Comment:
Again, I wasn't able to follow the whole discussion, but why can `None`/default value not be a stand in for "dynamically typed and initialized to None"?

I mean it's not the end of the world because they can say "declare as dynamically typed and default to NOT_SET" explicitly, but that just a bit tedious.

Again, not trying to say it should change, but usually in the "Alternatives" section of a REP you don't just say what you decided to do, but also why you decided that (summary of the development discussions), what you didn't choose, and why you didn't choose them. That could be helpful here too.

Reply:
> Again, I wasn't able to follow the whole discussion, but why can None/default value not be a stand in for "dynamically typed and initialized to None"?

I think I made this change to make people think if they really wanted a dynamically typed parameter, but not throwing an exception and instead making it dynamically typed also sounds reasonable to me (and maybe it will avoid breaking downstream packages).

Reply:
Deprecating this particular signature makes sense to me, since I think we want to encourage people to use statically typed parameters by default. Maybe we should document this rationale here.

Reply:
I'm fine with deprecating and then throwing, and therefore requiring explicitly opting into a dynamic parameter.

As I said in the other thread, I just think (based on past experience) people will want the more convenient way of getting dynamically typed parameters.

Reply:
> As I said in the other thread, I just think (based on past experience) people will want the more convenient way of getting dynamically typed parameters.

I think that could be added, something like:

```cpp
node->declare_parameter("parameter_name", rclcpp::PARAMETER_DYNAMIC);
node->declare_parameter("parameter_name", default_value, rclcpp::PARAMETER_DYNAMIC);
```

or maybe something to streamline the parameter descriptor construction:

```cpp
node->declare_parameter("parameter_name", default_value, rclcpp::ParameterDescriptor{}.dynamic_typing());
```

Reply:
Those look nicer.

Reply:
I've documented that as a possible improvement, see https://github.com/ros2/rclcpp/pull/1568/commits/46a773e5640648b7ffae96824233ea6f1c13e415